### PR TITLE
stable setState and improved types

### PR DESCRIPTION
**Pull request by KutnerUri**
- stable setState - wrap method with useCallback, to reuse the same value (mabye I should put `onChange` to be more consistent with `setState`?)
- improve types - usage is now forced to supply either `initialValue` or `controlledValue`, return type is Value, which also works when setting `initialValue: undefined` or `controlledValue: undefined`

### CODE REVIEW

**Comment by KutnerUri** [comment]
also - I don't like the limitation of "controlledValue must not be undefined".
undefined should be a valid value, not sure how to work around it

File: packages/use-optionally-controlled-state/src/index.tsx [hunk 1]
```diff
@@ -1,23 +1,31 @@
...
-  initialValue?: Value;
-  onChange?(value: Value): void;
-}): [Value | undefined, (value: Value) => void] {
+}: OptionalState<Value>): [Value, (value: Value) => void] {
   const isControlled = controlledValue !== undefined;
   const initialIsControlled = useConstant(() => isControlled);
   const [stateValue, setStateValue] = useState(initialValue);
 
-  if (__DEV__) {
+  if (typeof __DEV__ !== 'undefined' && __DEV__) {
```
Comment (amannn) [inline packages/use-optionally-controlled-state/src/index.tsx:28]: This is rewritten at build-time, please revert this change.

File: packages/use-optionally-controlled-state/src/index.tsx [hunk 2]
```diff
@@ -37,12 +45,17 @@ export default function useOptionallyControlledState<Value>({
...
 
-  const value = isControlled ? controlledValue : stateValue;
+  const value = isControlled ? controlledValue : (stateValue as Value);
 
-  function onValueChange(nextValue: Value) {
-    if (!isControlled) setStateValue(nextValue);
-    if (onChange) onChange(nextValue);
-  }
+  const isControlledRef = useRef(false);
+  isControlledRef.current = isControlled;
```
Comment (amannn) [inline packages/use-optionally-controlled-state/src/index.tsx:51]: I think this is not concurrent mode safe. Can we make `isControlled` a dependency of the callback? If `isControlled` changes, we throw an error anyway.

In the future we might want to use [`useEvent`](https://github.com/reactjs/rfcs/pull/220) here.

**Review by amannn — COMMENTED** [review]
Thanks a lot for your helpful contribution!

> also - I don't like the limitation of "controlledValue must not be undefined".
> undefined should be a valid value, not sure how to work around it

Do you mean this error?

> Can not change from controlled to uncontrolled mode. If `undefined` needs to be used for controlled values, please use `null` instead.

That's a similar check that React does for inputs. It's important to know if a component is consistently controlled or uncontrolled over its lifetime.

File: packages/use-optionally-controlled-state/src/index.tsx (hunk 1 above)
Comment (KutnerUri) [reply packages/use-optionally-controlled-state/src/index.tsx:28]: I don't know what is writing `__DEV__`. I would guess it assumes there is a Webpack `DefinePlugin` somewhere replacing the variable.
While this works out of the box in CRA, I think this is prone to bugs and will throw an `__DEV__ is not defined` error. So, I strongly suggest adding this safeguard.

if you still think it's unnecessary, I will remove it :)

File: packages/use-optionally-controlled-state/src/index.tsx (hunk 2 above)
Comment (KutnerUri) [reply packages/use-optionally-controlled-state/src/index.tsx:51]: hmm you are right. I wanted it to behave like useState (which provides a stable setState), but I didn't think about concurrency. Let me read into it

**Comment by KutnerUri** [comment]
I see. React will show these _warnings_ for controlled elements:

Passing `value` without `onChange`:
```
Warning: You provided a `value` prop to a form field without an `onChange` handler. This will render a read-only field. If the field should be mutable use `defaultValue`. Otherwise, set either `onChange` or `readOnly`.
```

changing `value` from `undefined` to defined: (also vise versa)
```
Warning: A component is changing a controlled input to be uncontrolled. This is likely caused by the value changing from a defined to undefined, which should not happen. Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://reactjs.org/link/controlled-components
```

This is sensible. I guess the weird part is that 'controlled' elements have a string value (which can't really be undefined), and this hook can be anything.

File: packages/use-optionally-controlled-state/src/index.tsx (hunk 2 above)
Comment (KutnerUri) [reply packages/use-optionally-controlled-state/src/index.tsx:51]: oh no, this is not safe at all. and I used it all over the place 😱 

so that's why `useLatest` used `useEffect` to update the current value. 🤔 
oh well. I will add it to the dep array

File: packages/use-optionally-controlled-state/src/index.tsx [hunk 3]
```diff
@@ -37,12 +45,17 @@ export default function useOptionallyControlledState<Value>({
...
+  const onValueChange = useCallback(
+    (nextValue: Value) => {
+      if (!isControlledRef.current) setStateValue(nextValue);
+      if (onChange) onChange(nextValue);
+    },
+    [onChange]
+  );
```
Comment (KutnerUri) [inline packages/use-optionally-controlled-state/src/index.tsx:58]:
```suggestion
  const onValueChange = useCallback(
    (nextValue: Value) => {
      if (!isControlled) setStateValue(nextValue);
      if (onChange) onChange(nextValue);
    },
    [onChange, isControlled]
  );
```

File: packages/use-optionally-controlled-state/src/index.tsx (hunk 1 above)
Comment (amannn) [reply packages/use-optionally-controlled-state/src/index.tsx:28]: There are separate dev and prod builds, this variable never makes it into the final bundle: https://unpkg.com/browse/use-optionally-controlled-state@1.1.0/dist/

So yes, please remove the check 🙂

File: packages/use-optionally-controlled-state/src/index.tsx [hunk 4]
```diff
@@ -1,23 +1,31 @@
-import {useState} from 'react';
+import {useState, useCallback, useRef} from 'react';
```
Comment (amannn) [inline packages/use-optionally-controlled-state/src/index.tsx:1]: Can you make sure `yarn lint` and `yarn test` passes? I didn't have time to set this up on CI yet.

The import of `useRef` should be flagged here, since this is not used.

File: packages/use-optionally-controlled-state/src/index.tsx [hunk 5]
```diff
@@ -1,23 +1,31 @@
...
 import useConstant from 'use-constant';
 
+export type OptionalState<Value> =
```
Comment (amannn) [inline packages/use-optionally-controlled-state/src/index.tsx:4]: I wouldn't export this, I think it's sufficient to have this as a private type. Mostly because I'm a bit unsure about the naming. I think it's not "optional state" but rather the "options" the hook takes.

File: packages/use-optionally-controlled-state/src/index.tsx [hunk 6]
```diff
@@ -1,23 +1,31 @@
...
   onChange
-}: {
-  controlledValue?: Value;
-  initialValue?: Value;
-  onChange?(value: Value): void;
-}): [Value | undefined, (value: Value) => void] {
+}: OptionalState<Value>): [Value, (value: Value) => void] {
   const isControlled = controlledValue !== undefined;
   const initialIsControlled = useConstant(() => isControlled);
   const [stateValue, setStateValue] = useState(initialValue);
```
Comment (amannn) [inline packages/use-optionally-controlled-state/src/index.tsx:26]: Maybe we could add the `Value` generic here to avoid the cast in L48?

File: packages/use-optionally-controlled-state/src/index.tsx (hunk 5 above)
Comment (KutnerUri) [reply packages/use-optionally-controlled-state/src/index.tsx:4]: yea but OptionalStateOptions was a bit too weird haha

File: packages/use-optionally-controlled-state/src/index.tsx (hunk 1 above)
Comment (KutnerUri) [reply packages/use-optionally-controlled-state/src/index.tsx:28]: ooh, right

File: packages/use-optionally-controlled-state/src/index.tsx [hunk 7]
```diff
@@ -37,12 +45,16 @@ export default function useOptionallyControlledState<Value>({
     }
   }
 
-  const value = isControlled ? controlledValue : stateValue;
+  // Options type ensures that at either controlledValue or stateValue has a value
```
Comment (amannn) [inline packages/use-optionally-controlled-state/src/index.tsx:48]:
```suggestion
  // Options type ensures that either `controlledValue` or `stateValue` is defined
```

File: packages/use-optionally-controlled-state/src/index.tsx (hunk 6 above)
Comment (KutnerUri) [reply packages/use-optionally-controlled-state/src/index.tsx:26]: no, initialValue is optional, I'd rather only cast it when needed. 
The whole idea was to make sure the value doesn't get `| undefined` in the consumer, like:
```
const [value, setValue] = useOptionalState({ initialValue: 1 });
// value was `number | undefined`
```

The new `Options` type enforces that the user specifies either `initialValue` or `controlledValue`.

On the plus side, I replaced the casting with the `!` operator, verified these work:
```tsx
function typeTest(maybe?: number) {
  const [value1, setValue1] = useOptionallyControlledState({ controlledValue: 1 }); // type is number
  const [value2, setValue2] = useOptionallyControlledState({ initialValue: 1 });  // type is number
  const [value3, setValue3] = useOptionallyControlledState({ initialValue: undefined });  // type is undefined
  const [value4, setValue4] = useOptionallyControlledState({ initialValue: undefined });  // type is undefined
  const [value5, setValue5] = useOptionallyControlledState({ initialValue: maybe });  // type is number | undefined
}
```

actually, we can enforce the types not to allow both initialValue and controlledValue, but I think it will cause bugs than value

File: packages/use-optionally-controlled-state/src/index.tsx (hunk 1 above)
Comment (KutnerUri) [reply packages/use-optionally-controlled-state/src/index.tsx]: removed

File: packages/use-optionally-controlled-state/src/index.tsx (hunk 4 above)
Comment (KutnerUri) [reply packages/use-optionally-controlled-state/src/index.tsx]: verified with yarn lint 👍

**Comment by KutnerUri** [comment]
I verified `npm run lint` and `npm run build` pass 😃

**Comment by amannn** [comment]
Published in `use-optionally-controlled-state@1.2.0` – many thanks!
